### PR TITLE
Support quoted identifiers in `SimpleJdbcInsert`

### PR DESCRIPTION
**Pull request by kitaisreal**
Motivation https://github.com/spring-projects/spring-framework/issues/13874

**Current implementation steps**
1.  Getting identifierQuoteString to use for escaping from dataBaseMetadata in `GenericTableMetadataProvider` http://docs.oracle.com/javase/6/docs/api/java/sql/DatabaseMetaData.html#getIdentifierQuoteString%28%29.
2. Added property usingEscaping in `SimpleJdbcInsert` and logic related to handling it in `TableMetadataContext`.
3. Updated tests.

### CODE REVIEW

**Comment by Dretch** [comment]
This is awesome, thank you very much!

**Comment by foldvari** [comment]
As far as I see this solution does not support cases when the opening and the closing character used for quoting is different. (E.g. MS SQL [identifier] ).

File: spring-jdbc/src/main/java/org/springframework/jdbc/core/metadata/GenericTableMetaDataProvider.java [hunk 1]
```diff
@@ -76,6 +76,8 @@
 	/** Collection of TableParameterMetaData objects. */
 	private List<TableParameterMetaData> tableParameterMetaData = new ArrayList<>();
 
+	/** the string used to quote SQL identifiers. */
+	private String identifierQuoteString = "";
```
Comment (foldvari) [inline spring-jdbc/src/main/java/org/springframework/jdbc/core/metadata/GenericTableMetaDataProvider.java]: There is at least one RDBMS (MS SQL) which uses different characters for opening and closing. ( [identifier] )

Comment (kitaisreal) [reply spring-jdbc/src/main/java/org/springframework/jdbc/core/metadata/GenericTableMetaDataProvider.java]: Could you please provide link to documentation where it is mentioned ? I could only find this https://docs.microsoft.com/en-us/sql/odbc/reference/develop-app/quoted-identifiers?view=sql-server-ver15. 
We are getting this identifier to quote SQL from database metadata https://docs.oracle.com/javase/6/docs/api/java/sql/DatabaseMetaData.html#getIdentifierQuoteString%28%29. If this class does not support case where there are 2 different characters to quote SQL identifiers, we also could not.

Comment (foldvari) [reply spring-jdbc/src/main/java/org/springframework/jdbc/core/metadata/GenericTableMetaDataProvider.java]: Thanks for the links to the documentation. It seems that the driver takes care about switching on the support of quoted identifiers (QUOTED_IDENTIFIER) so the change will work with MS SQL as well.

**Comment by manicmonkey** [comment]
@sbrannen you recently made changes in this area - can you comment on the viability of this PR?

**Review by sbrannen — CHANGES_REQUESTED** [review]
@kitaisreal,

Thanks for the PR, and sorry for taking so long to review this.

Do you have time to make the requested changes (consistently referring to "quote identifiers" instead of "using escaping") and rebase on `main`?

If not, just let us know, and we'll take over the PR.

Cheers,

Sam

**Comment by sbrannen** [comment]
> Do you have time to make the requested changes (consistently referring to "quote identifiers" instead of "using escaping") and rebase on `main`?

@kitaisreal, you can disregard that. I am taking care of that and other changes locally after merging your PR.

File: spring-jdbc/src/test/java/org/springframework/jdbc/core/simple/SimpleJdbcInsertTests.java [hunk 2]
```diff
@@ -81,4 +82,36 @@ public void testNoSuchTable() throws Exception {
...
+	}
+
+	@Test
+	public void testSimpleJdbcInsertWithEscapingWithSchemaName() throws Exception {
+		SimpleJdbcInsert jdbcInsert = new SimpleJdbcInsert(dataSource).withSchemaName("S").withTableName("T").usingColumns("F", "S").usingEscaping(true);
+
+		given(databaseMetaData.getIdentifierQuoteString()).willReturn("`");
+
+		jdbcInsert.compile();
+		String expected = "INSERT INTO `S.T` (`F`, `S`) VALUES(?, ?)";
```
Comment (sbrannen) [inline spring-jdbc/src/test/java/org/springframework/jdbc/core/simple/SimpleJdbcInsertTests.java:101]:
```suggestion
		String expected = "INSERT INTO `S`.`T` (`F`, `S`) VALUES(?, ?)";
```

Just a side note: the schema and table names have to be quoted independently.

I've fixed this in my local branch and added integration tests with H2 to verify it.

**Comment by sbrannen** [comment]
This has been merged into `main` in d39034754fc99862fc80eb8bd860004afd4b6d5a and revised in 7dc0653f38a0ac653bf29068dbc4bca632530eea. See also 070590cb1162ff5ef3cd2a3f7a64390caa58782c.

Thanks, and congratulations on getting your first PR for the Spring Framework merged! 👍

**Pull request closed without merging**